### ISSUE TITLE
broken --remove-output / --print-status

### ISSUE THREAD

**Issue by maadcoen** [columnflow/columnflow#520]
I noticed the output removal and status printing via the command line doesn't work properly anymore. The cf.SelectEvents task is sometimes indicated a incomplete while it isn't. That's also recognised by columnflow that doens't rerun the task despite the output of --print-status. When trying to remove the output of via --remove-output of the cf.SelectEvents task, nothing is removed.

I've tracked down the issue to the recent addition of optional histograms at the selection state. Since these are optional, columnflow considers the SelectEvents task finished if only the hists.pickle file is missing. This seems to work well when running the workflow, but I didn't check why that works. When printing the status, I found it `TargetCollection.status_text` gets called, which calls `TargetCollection.count`, which calls `TargetCollection._iter_state`. When removing files, `SiblingFileCollectionBase.remove` gets called, which calls `TargetCollection.iter_existing` which again calls  `TargetCollection._iter_state`. 

Now,   `TargetCollection._iter_state` iterates over all the groups of files that are complete: that is, if columns, stats, results, and hists are all present. It can take into account optional files, like "hists" but it only does so if one passes it `optional_existing=True`. One can pass this argument to `count` and `iter_existing`, but the problem is that this isn't done by neither `status_text` not `remove`.

**Comment by pkausw** [comment]
Thanks for creating the issue! I'm tagging @riga here since these features are mainly part of law

**Comment by maadcoen** [comment]
also, just noticed cf.MergeSelectionStats is affected in the same way, also because of the optional hists

**Comment by hephysicist** [comment]
Hi @pkausw @maadcoen @riga 
I have also encountered this weird behaviour of the cf.PlotVariables1D task after the update of the columnflow module up to the latest version. Now if I want to remove the output of the task and some predecessors, I get the following output:
```
[stzakhar@naf-cms12 scripts]$ ./03_plot_variables.sh run3 --remove-output 3
law run cf.PlotVariables1D --config run3_2022_preEE_tau_spinner --processes h_ggf_htt --datasets h_ggf_htt_filtered --version signal_test --categories incl --cf.CalibrateEvents-workflow htcondor --cf.SelectEvents-workflow htcondor --cf.ReduceEvents-workflow htcondor --cf.MergeReducedEvents-workflow htcondor --variables puppi_met_pt,muon_pt_1bin,tau_pt_1bin,hcand_invm_1bin --general-settings cms-label=pw --remove-output 3
remove task output with max_depth 3
removal mode? [i*(interactive), d(dry), a(all)] i
selected interactive mode

0 > cf.PlotVariables1D(effective_workflow=local, branch=-1, analysis=httcp.config.analysis_httcp.analysis_httcp, version=signal_test, config=run3_2022_preEE_tau_spinner, shift=nominal, local_shift=nominal, calib
    rators=main, selector=main, producers=main, categories=incl, variables=puppi_met_pt,muon_pt_1bin,tau_pt_1bin,hcand_invm_1bin, datasets=h_ggf_htt_filtered, processes=h_ggf_htt, weight_producer=all_weights, hi
    st_hooks=, file_types=pdf, plot_suffix=NO_STR, workflow=local)
      remove outputs? [y*(yes), n(no), a(all)] a
      collection: SiblingFileCollection(len=4, threshold=4.0, fs=local_fs, dir=$CF_STORE_LOCAL/analysis_httcp/cf.PlotVariables1D/run3_2022_preEE_tau_spinner/calib__main/sel__main/prod__main/weight__all_weights/n
                  ominal/datasets_h_ggf_htt_filtered/signal_test)
        removed
[stzakhar@naf-cms12 scripts]$ 
```
So looks like this task lost track of the previous tasks. But if I run a previous task (cf.MergeHistograms) and try to remove the outputs, it works as entended. 
```
[stzakhar@naf-cms12 scripts]$ law run cf.MergeHistograms --config run3_2022_preEE_tau_spinner --version signal_test --remove-output 3
remove task output with max_depth 3
removal mode? [i*(interactive), d(dry), a(all)] i
selected interactive mode

0 > cf.MergeHistograms(effective_workflow=local, branch=-1, analysis=httcp.config.analysis_httcp.analysis_httcp, version=signal_test, config=run3_2022_preEE_tau_spinner, shift=nominal, local_shift=nominal, datas
│   et=h_ggf_htt_filtered, calibrators=main, selector=main, producers=main, variables=event, weight_producer=all_weights, only_missing=False, workflow=local)
│     remove outputs? [y*(yes), n(no), a(all)] n
│
└──1 > cf.CreateHistograms(effective_workflow=local, branch=-1, analysis=httcp.config.analysis_httcp.analysis_httcp, version=signal_test, config=run3_2022_preEE_tau_spinner, shift=nominal, local_shift=nominal, d
   │   ataset=h_ggf_htt_filtered, calibrators=main, selector=main, producers=main, variables=event, weight_producer=all_weights, workflow=local)
   │     remove outputs? [y*(yes), n(no), a(all)] n
   │
   ├──2 > cf.ProvideReducedEvents(effective_workflow=local, branch=-1, analysis=httcp.config.analysis_httcp.analysis_httcp, version=signal_test, config=run3_2022_preEE_tau_spinner, shift=nominal, local_shift=nom
   │  │   inal, dataset=h_ggf_htt_filtered, calibrators=main, selector=main, skip_merging=False, force_merging=False, workflow=local)
   │  │     remove outputs? [y*(yes), n(no), a(all)] n
   │  │
   │  ├──3 > cf.MergeReductionStats(analysis=httcp.config.analysis_httcp.analysis_httcp, version=signal_test, config=run3_2022_preEE_tau_spinner, shift=nominal, local_shift=nominal, dataset=h_ggf_htt_filtered, c
   │  │      alibrators=main, selector=main, n_inputs=10)
   │  │        remove outputs? [y*(yes), n(no), a(all)] n
   │  │
   │  └──3 > cf.MergeReducedEvents(effective_workflow=local, branch=0, tree_index=-1, tree_depth=0, analysis=httcp.config.analysis_httcp.analysis_httcp, version=signal_test, config=run3_2022_preEE_tau_spinner, s
   │         hift=nominal, local_shift=nominal, dataset=h_ggf_htt_filtered, calibrators=main, selector=main)
   │           remove outputs? [y*(yes), n(no), a(all)] n
   │
   └──2 > cf.ProduceColumns(effective_workflow=local, branch=-1, analysis=httcp.config.analysis_httcp.analysis_httcp, version=signal_test, config=run3_2022_preEE_tau_spinner, shift=nominal, local_shift=nominal, 
      │   dataset=h_ggf_htt_filtered, calibrators=main, selector=main, producer=main, workflow=local)
      │     remove outputs? [y*(yes), n(no), a(all)] n
      │
      ├──3 > cf.ProvideReducedEvents(effective_workflow=local, branch=-1, analysis=httcp.config.analysis_httcp.analysis_httcp, version=signal_test, config=run3_2022_preEE_tau_spinner, shift=nominal, local_shift=
      │      nominal, dataset=h_ggf_htt_filtered, calibrators=main, selector=main, skip_merging=False, force_merging=False, workflow=local)
      │        already handled
      │
      ├──3 > cf.MergeSelectionStats(effective_workflow=local, branch=-1, tree_index=0, tree_depth=0, analysis=httcp.config.analysis_httcp.analysis_httcp, version=signal_test, config=run3_2022_preEE_tau_spinner, 
      │      shift=nominal, local_shift=nominal, dataset=h_ggf_htt_filtered, calibrators=main, selector=main, workflow=local)
      │        remove outputs? [y*(yes), n(no), a(all)] n
      │
      └──3 > cf.BundleExternalFiles(source_path=NO_STR, analysis=httcp.config.analysis_httcp.analysis_httcp, config=run3_2022_preEE_tau_spinner, replicas=5)
               remove outputs? [y*(yes), n(no), a(all)] n
[stzakhar@naf-cms12 scripts]$ 
```
If you need any additional information, I'm willing to provide it to you! 

Best regards, 
Stepan.

**Comment by mafrahm** [comment]
Hi @maadcoen, thank you for bringing that to our attention. I was able to reproduce the issue. It looks like the `optional` tag of the output does not work as intended in all cases but I have yet to figure out why this is the case.

@hephysicist I think your issue is a separate problem that is caused by these two lines:  https://github.com/columnflow/columnflow/blob/master/columnflow/tasks/plotting.py#L217-L218
Removing these lines should fix your problem, however I am not sure why they were added in the first place. Maybe @riga  can provide more details.
